Pt 2247

Tusamarco opened the pull request:
- [ x] The contributed code is licensed under GPL v2.0
- [ x] Contributor Licence Agreement (CLA) is signed
- [ NA] util/update-modules has been ran
- [ x] Documentation updated
- [NA ] Test suite update

@svetasmirnova please check and review also https://jira.percona.com/browse/PT-2247

Tusamarco committed:
I had to correct few things in the script to have it working correctly:
1) OR instead and in the if line 2042
2) evaluate array create_user if it has some value
3) filter out the user creation in Mariadb inside the Grants

Tusamarco committed:
Fixing other small things and MariaDB specific,
- Use of VIA
- Use of USING
- Incompatible syntax with MySQL
- ADD IF NOT EXISTS
- convert the script to us AS to maintain the hash as they are

Add a parameter --convert-MariaDB as parameter (default false)

Tusamarco committed:
Fixed one issue (I introduced filtering the PASSWORD keyword)
correct an IF condition
removed a delete action on mysql.user. Always use DROP not delete

Tusamarco committed:
Significant refactor of the create in case of MySQL 8 and caching_sha2_password plugin.
The problem is that when caching_sha2_password is used the character utilised in the string may be invalid for MySQL itself to process during creation time.
Given that the password must be converted to HEX and then pushed as binary using the AS <bin password> format

Tusamarco committed:
small bug in regex

Tusamarco committed:
Adding replication safe condition for drop

Tusamarco committed:
Merge pull request #685 from Tusamarco/PT-2247-1212

Pt 2247 1212

Tusamarco committed:
Adding the option to use in session print_identified_with_as_hex as an alternative to the conversion
Param --print_identified_with_as_hex

Tusamarco committed:
Merge pull request #687 from Tusamarco/PT-2247-1212

Adding the option to use in session print_identified_with_as_hex as a…

Tusamarco committed:
Adding a small check if we have MariaDB in order to prevent usage of print_identified_with_as_hex

Tusamarco committed:
Merge pull request #688 from Tusamarco/PT-2247-1212

Adding a small check if we have MariaDB in order to prevent usage of …

svetasmirnova reviewed (changes requested):
PLease add test case. You can find how in the contribution guide and examples in t/pt-show-grants/

svetasmirnova commented on bin/pt-show-grants:
```diff
@@ -2039,25 +2039,52 @@ sub main {
       # If MySQL 5.7.6+ then we need to use SHOW CREATE USER
       my @create_user;
-      if (( VersionCompare::cmp($version, '5.7.6') >= 0 ) &&
+      if (( VersionCompare::cmp($version, '5.7.6') >= 0 ) ||
```
With "||" this expression is always true.

svetasmirnova commented on bin/pt-show-grants:
```diff
@@ -2433,6 +2467,16 @@ type: array
 
+=item --convert-MariaDB
+
+When set it convert some of the proprietary MariaDB syntax into valid MySQL form
```
Can't we do this automatically if we connect to MariaDB? Or will it break for valid MySQL syntax?

svetasmirnova replied:
Looks like not, useful option.

svetasmirnova commented on bin/pt-show-grants:
```diff
@@ -2433,6 +2467,16 @@ type: array
+
+When set it convert some of the proprietary MariaDB syntax into valid MySQL form
+
+=item --print_identified_with_as_hex
```
According to the user manual at https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_print_identified_with_as_hex:

> Hash values that do not contain unprintable characters still display as regular string literals, even with this variable enabled. 

Do we really need this option or checking for correct MySQL version is enough?

svetasmirnova committed:
Merge branch '3.x' into pt-2247

svetasmirnova committed:
PT-2247 - pt-show-grants does not CREATE USER

- Removed option print_identified_with_as_hex, because it was already
  implemented in the PT-2190 fix
- Simplified patch
- Kept CREATE USER/ALTER USER sequence and extra DELETE from mysql.user
  table
- Added test case

svetasmirnova commented:
@Tusamarco I changed your PR: simplified a few things, removed the option --print_identified_with_hex (made it default and only behavior), and added a test case. Please review and see if you agree with the changes.

Tusamarco reviewed (commented):
LGTM
